### ISSUE TITLE
Daily update stuck at importing way

### ISSUE THREAD

**Issue by rcarroll0452** [openstreetmap/Nominatim#1550]
I just completed a full planet import and commenced daily updates. But it seems to be stuck at processing "way", similar to this #426 

It has been at `Processing: Node(2558k 19.1k/s) Way(64k 1.52k/s) Relation(0 0.00/s)` for more than 6 hours now.

pg_stat_activity:
```
 datid |  datname  | pid  | usesysid |  usename  | application_name | client_addr | client_hostname | client_port |          backend_start           |            xact_start            |           query_start            |           state_change           | waiting | state  | backend_xid | backend_xmin |                                                                    query                                                                    
-------+-----------+------+----------+-----------+------------------+-------------+-----------------+-------------+----------------------------------+----------------------------------+----------------------------------+----------------------------------+---------+--------+-------------+--------------+---------------------------------------------------------------------------------------------------------------------------------------------
 36763 | nominatim | 1229 |    16384 | nominatim |                  |             |                 |          -1 | 2019-11-04 13:22:42.526633+05:30 |                                  | 2019-11-04 13:22:42.534614+05:30 | 2019-11-04 13:22:42.53465+05:30  | f       | idle   |             |              | DEALLOCATE pdo_stmt_00000001
 36763 | nominatim | 1407 |    16384 | nominatim | osm2pgsql        |             |                 |          -1 | 2019-11-04 13:23:54.500129+05:30 | 2019-11-04 13:26:55.410737+05:30 | 2019-11-04 13:26:55.410737+05:30 | 2019-11-04 13:26:55.41074+05:30  | f       | active |             |    262556575 | COPY planet_osm_ways FROM STDIN
 36763 | nominatim | 1408 |    16384 | nominatim | osm2pgsql        |             |                 |          -1 | 2019-11-04 13:23:54.502136+05:30 |                                  | 2019-11-04 13:26:51.355078+05:30 | 2019-11-04 13:26:51.355416+05:30 | f       | idle   |             |              | PREPARE mark_rels_by_node(int8) AS select id from planet_osm_ways WHERE nodes && ARRAY[$1];                                                +
       |           |      |          |           |                  |             |                 |             |                                  |                                  |                                  |                                  |         |        |             |              | PREPARE mark_rels_by_way(int8) AS select id from planet_osm_rels WHERE parts && ARRAY[$1] AND parts[way_off+1:rel_off] && ARRAY[$1];       +
       |           |      |          |           |                  |             |                 |             |                                  |                                  |                                  |                                  |         |        |             |              | PREPARE mark_rels(int8) AS select id from planet_osm_rels WHERE parts && ARRAY[$1] AND parts[rel_off+1:array_length(parts,1)] && ARRAY[$1];+
       |           |      |          |           |                  |             |                 |             |                                  |                                  |                                  |                                  |         |        |             |              | 
 36763 | nominatim | 1409 |    16384 | nominatim | osm2pgsql        |             |                 |          -1 | 2019-11-04 13:23:54.507759+05:30 |                                  | 2019-11-04 13:23:54.511132+05:30 | 2019-11-04 13:23:54.511679+05:30 | f       | idle   |             |              | PREPARE get_rel (int8) AS SELECT members, tags, array_upper(members,1)/2 FROM planet_osm_rels WHERE id = $1;                               +
       |           |      |          |           |                  |             |                 |             |                                  |                                  |                                  |                                  |         |        |             |              | PREPARE rels_using_way(int8) AS SELECT id FROM planet_osm_rels WHERE parts && ARRAY[$1] AND parts[way_off+1:rel_off] && ARRAY[$1];         +
       |           |      |          |           |                  |             |                 |             |                                  |                                  |                                  |                                  |         |        |             |              | 
 36763 | nominatim | 1411 |    16384 | nominatim | osm2pgsql        |             |                 |          -1 | 2019-11-04 13:23:54.512719+05:30 | 2019-11-04 13:25:30.93609+05:30  | 2019-11-04 13:25:30.93609+05:30  | 2019-11-04 13:25:30.936093+05:30 | f       | active |   262556575 |    262556575 | COPY place FROM STDIN
 36763 | nominatim | 1412 |    16384 | nominatim | osm2pgsql        |             |                 |          -1 | 2019-11-04 13:23:54.514572+05:30 | 2019-11-04 13:26:51.355625+05:30 | 2019-11-04 13:26:51.355625+05:30 | 2019-11-04 13:26:51.355626+05:30 | t       | active |   262557705 |    262556575 | DELETE FROM place WHERE osm_type = 'W' AND osm_id = 257126013 and class = any(ARRAY['highway'])
 36763 | nominatim | 7846 |    16384 | nominatim | psql             |             |                 |          -1 | 2019-11-04 16:32:55.221453+05:30 | 2019-11-04 17:22:12.249178+05:30 | 2019-11-04 17:22:12.249178+05:30 | 2019-11-04 17:22:12.249184+05:30 | f       | active |             |    262556575 | select * from pg_stat_activity;
```

I believe this to be the deadlock:
```
 blocked_pid | blocked_user | blocking_pid | blocking_user |                                        blocked_statement                                        | current_statement_in_blocking_process 
-------------+--------------+--------------+---------------+-------------------------------------------------------------------------------------------------+---------------------------------------
        1412 | nominatim    |         1411 | nominatim     | DELETE FROM place WHERE osm_type = 'W' AND osm_id = 257126013 and class = any(ARRAY['highway']) | COPY place FROM STDIN
```

**Comment by lonvia** [comment]
Might be related to https://github.com/openstreetmap/osm2pgsql/issues/964

**Comment by mmd-osm** [comment]
Rather than a deadlock, I believe this is a livelock. We have two processes doing both synchronous and asynchronous database updates on the same table at the same time:

* Pid 1411: asynchronous processing of queue entries via db_copy_thread_t, 
currently processing `COPY place FROM STDIN`  (not waiting)

+ Pid 1412: synchronous deleting of place entries via output_gazetteer_t::delete_unused_classes, 
currently waiting in `DELETE FROM place WHERE osm_type = 'W' AND osm_id = 257126013 and class = any(ARRAY['highway'])`

It seems that Pid 1411 is holding an exclusive lock on table _place_ with osm_type = 'W' AND osm_id = 257126013, while processing queue entries. Pid 1412 gets stuck on the DELETE, as it can't acquire a lock on this entry. Question: HOW/WHY does this happen?

At this point, the whole process gets stuck, as output_gazetteer_t::process_* can't make any progress, and the condition variable on _m_queue_cond.wait_ doesn't use a timeout to eventually abort the whole process ( _std::condition_variable::wait_for_  rather than _wait_ would be required to abort waiting).

**Comment by txtor** [comment]
Same problem here. Updating Europe stucks
```Processing: Node(807k 13.2k/s) Way(32k 0.60k/s) Relation(0 0.00/s)```

blocked query: `DELETE FROM place WHERE osm_type = 'W' AND osm_id = 251326812 and class = any(ARRAY['highway'])`
blocking query: ` COPY place FROM STDIN`

osm2pgsql version 1.2.1 (64 bit id space)
Nominatim 3.4.0
psql (PostgreSQL) 10.11

**Comment by garymg18** [comment]
Same for me (full update): 

```bash
docker exec -it nominatim sudo -u postgres ./src/build/utils/update.php --import-osmosis-all
Currently at sequence 2605 (30/10/2019 23:58:57) - 1 indexed
/usr/local/bin/pyosmium-get-changes --server https://ftp5.gwdg.de/pub/misc/openstreetmap/planet.openstreetmap.org/replication/day/ -o /app/src/build/osmosischange.osc -s 30 -I 2605
/app/src/build/osm2pgsql/osm2pgsql -klas --number-processes 1 -C 2000 -O gazetteer -S /app/src/settings/import-full.style -d nominatim -P 5432 /app/src/build/osmosischange.osc
osm2pgsql version 1.2.1 (64 bit id space)

Allocating memory for dense node cache
Allocating dense node cache in one big chunk
Allocating memory for sparse node cache
Sharing dense sparse
Node-cache: cache=2000MB, maxblocks=32000*65536, allocation method=11
Mid: pgsql, cache=2000
Parsing gazetteer style file '/app/src/settings/import-full.style'.
Using projection SRS 4326 (Latlong)

Reading in file: /app/src/build/osmosischange.osc
Using XML parser.
Processing: Node(3049k 41.8k/s) Way(37k 1.16k/s) Relation(0 0.00/s)
```

Postgres: 

```
nominatim=# SELECT blocked_locks.pid     AS blocked_pid,
         blocked_activity.usename  AS blocked_user,
         blocking_locks.pid     AS blocking_pid,
         blocking_activity.usename AS blocking_user,
         blocked_activity.query    AS blocked_statement,
         blocking_activity.query   AS current_statement_in_blocking_process
   FROM  pg_catalog.pg_locks         blocked_locks
    JOIN pg_catalog.pg_stat_activity blocked_activity  ON blocked_activity.pid = blocked_locks.pid
    JOIN pg_catalog.pg_locks         blocking_locks
        ON blocking_locks.locktype = blocked_locks.locktype
        AND blocking_locks.DATABASE IS NOT DISTINCT FROM blocked_locks.DATABASE
        AND blocking_locks.relation IS NOT DISTINCT FROM blocked_locks.relation
        AND blocking_locks.page IS NOT DISTINCT FROM blocked_locks.page
        AND blocking_locks.tuple IS NOT DISTINCT FROM blocked_locks.tuple
        AND blocking_locks.virtualxid IS NOT DISTINCT FROM blocked_locks.virtualxid
        AND blocking_locks.transactionid IS NOT DISTINCT FROM blocked_locks.transactionid
        AND blocking_locks.classid IS NOT DISTINCT FROM blocked_locks.classid
        AND blocking_locks.objid IS NOT DISTINCT FROM blocked_locks.objid
        AND blocking_locks.objsubid IS NOT DISTINCT FROM blocked_locks.objsubid
        AND blocking_locks.pid != blocked_locks.pid

    JOIN pg_catalog.pg_stat_activity blocking_activity ON blocking_activity.pid = blocking_locks.pid
   WHERE NOT blocked_locks.GRANTED;
 blocked_pid | blocked_user | blocking_pid | blocking_user |                                        blocked_statement                                        | current_statement_in_blocking_process
-------------+--------------+--------------+---------------+-------------------------------------------------------------------------------------------------+---------------------------------------
         364 | postgres     |          363 | postgres      | DELETE FROM place WHERE osm_type = 'W' AND osm_id = 122662475 and class = any(ARRAY['highway']) | COPY place FROM STDIN
(1 row)

nominatim=#
```

What is interesting, I managed to fully update other node with osm2pgsql version 0.96.0 (64 bit id space) installed from apt repo

**Comment by garymg18** [comment]
Ok, I managed to move on. @txtor I don't know how You did get version 1.2.1 of osm2pgsql, I just simply cloned osm2pgsql repo and switched to tag 1.2.1 - and that was the mistake. The fix is merged to master AFAICS. So simply clone the master branch and rebuild the osm2pgsql as here: https://github.com/openstreetmap/Nominatim/issues/1598#issuecomment-564120275 (v 1.2.0 'works')

And now I have the segmentation fault problem, but I saw that the problem is known and the task is opened

```
docker exec -it nominatim sudo -u postgres ./src/build/utils/update.php --import-osmosis-all
Currently at sequence 2605 (30/10/2019 23:58:57) - 1 indexed
/usr/local/bin/pyosmium-get-changes --server https://ftp5.gwdg.de/pub/misc/openstreetmap/planet.openstreetmap.org/replication/day/ -o /app/src/build/osmosischange.osc -s 30 -I 2605
/app/src/build/osm2pgsql/osm2pgsql -klas --number-processes 1 -C 2000 -O gazetteer -S /app/src/settings/import-full.style -d nominatim -P 5432 /app/src/build/osmosischange.osc
osm2pgsql version 1.2.0 (64 bit id space)

Allocating memory for dense node cache
Allocating dense node cache in one big chunk
Allocating memory for sparse node cache
Sharing dense sparse
Node-cache: cache=2000MB, maxblocks=32000*65536, allocation method=11
Mid: pgsql, cache=2000
Parsing gazetteer style file '/app/src/settings/import-full.style'.
Using projection SRS 4326 (Latlong)

Reading in file: /app/src/build/osmosischange.osc
Using XML parser.
Processing: Node(3049k 50.8k/s) Way(142k 2.37k/s) Relation(0 0.00/s)Segmentation fault
Error executing osm2pgsql: 139
```

https://github.com/openstreetmap/Nominatim/issues/1600

**Comment by txtor** [comment]
thanks @garymg18

**Comment by lonvia** [comment]
Fixed in bc68ff1e43f2527039640b8642bbe533814a3561.

**Comment by MatthiasLohr** [comment]
Is this worth a patch release?
